### PR TITLE
Create iso files only if Disk is present

### PR DESCRIPTION
**Pull request by germag**
**What this PR does / why we need it**:

Currently, using a virtiofs filesystem you can share ConfigMaps, Secrets and DownwardAPI volumes with the VM.  And if we share those volumes using virtiofs all the iso files are created since some parts of the code assumes that those volumes always have a Disk associated.

This PR refactors the code, so the iso files are not created if it's not necessary. It also removes the duplication of source/iso disk path calculation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This is PR was part of #9254, but as requested we only included the refactoring commits

**Release note**:
-->
```release-note
NONE
```

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -182,3 +186,49 @@ func findIsoSize(vmi *v1.VirtualMachineInstance, volume *v1.Volume, emptyIso boo
...
+	for _, volume := range vmi.Spec.Volumes {
+		if info.isValidType(&volume) {
+			volumes[volume.Name] = volume
+		}
+	}
+
+	for _, disk := range vmi.Spec.Domain.Devices.Disks {
+		volume, ok := volumes[disk.Name]
+		if !ok {
+			log.DefaultLogger().V(2).Warningf("No matching volume with name %s found", disk.Name)
```
Comment (vasiliy-ul) [inline pkg/config/config.go:208]: I think this warning will be a bit misleading. Here, `volumes` map is populated only with volumes of a valid type `info.isValidType`. While the iteration through the disks does not take the type into account. So a disk may in fact have a matching volume, but of a different type. I would drop the warning.

File: pkg/config/config-map_test.go [hunk 2]
```diff
@@ -64,11 +64,33 @@ var _ = Describe("ConfigMap", func() {
...
+						Name: "test-config",
+					},
+				},
+			},
+		})
+
+		err := CreateConfigMapDisks(vmi, false)
+		Expect(err).NotTo(HaveOccurred())
+		_, err = os.Stat(filepath.Join(ConfigMapDisksDir, "configmap-volume.iso"))
+		Expect(err).To(HaveOccurred())
```
Comment (vasiliy-ul) [inline pkg/config/config-map_test.go:93]: Probably better to be more explicit here:
```suggestion
		Expect(err).To(MatchError(os.ErrNotExist))
```
Same with other tests.

Comment (alicefr) [reply pkg/config/config-map_test.go:93]: Or maybe we should simply check that the directory is empty or has the config map file inside? I know that we are switching from having iso to files to directly using the files, but for the test itself I think it should test the new behavior in general and not verify that the old one still applies.

Comment (germag) [reply pkg/config/config-map_test.go:93]: > Or maybe we should simply check that the directory is empty or has the config map file inside? I know that we are switching from having iso to files to directly using the files, but for the test itself I think it should test the new behavior in general and not verify that the old one still applies.

I'm not sure to understand, we're keeping the old behavior of generating iso files, just with the restriction of creating them only if a Disk device exists. Or you are suggesting that we should also check if the ConfigMap volume was created correctly?

Comment (alicefr) [reply pkg/config/config-map_test.go:93]: It was not very clear, let me rephrase.
My comment was more about the purpose of the test. This test now checks that the old behavior doesn't apply, in other words, checking that the iso doesn't exist in the path. 
I understand the reasoning behind it, but I think it would make more sense if we check in the test, that the new behavior applies instead of checking that the old one doesn'appöy anymore. For example, if you expect an empty directory, then check that the directory is empty. 
I hope I manage to express it in a better way, let me know

Comment (germag) [reply pkg/config/config-map_test.go:93]: Yes, thanks for the clarification. I'll change the tests.

**Comment by germag** [comment]
v2:
- Remove misleading warning
- Check for empty disk dir in tests

**Comment by alicefr** [comment]
Looks good to me!
/lgtm

**Review by xpivarc — COMMENTED** [review]
/approve
/hold

Holding just to give a time to response.

File: pkg/config/config.go [hunk 3]
```diff
@@ -182,3 +184,48 @@ func findIsoSize(vmi *v1.VirtualMachineInstance, volume *v1.Volume, emptyIso boo
...
 }
+
+type volumeInfo interface {
+	isValidType(*v1.Volume) bool
+	getSourcePath(*v1.Volume) string
+	getIsoPath(*v1.Volume) string
+	getLabel(*v1.Volume) string
+}
+
+func createIsoDisksForConfigVolumes(vmi *v1.VirtualMachineInstance, emptyIso bool, info volumeInfo) error {
```
[the remark this answers is not in the document]
Comment (germag) [reply pkg/config/config.go:195]: there are a couple of things that I don't like of that api (I think it was my first approach):
- it's not obvious (so error-prone) that the caller _must_  pass volumes already filtered to createIsoDisksForConfigVolumes()
- sourceAndTargetPath() returning a tuple I think is error-prone, you need to check which one go first/second the source/target.

I like that you can pass a lambda expression, tho'

[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/config/config.go:195]: Btw, feel free to remove the hold (/hold cancel).

>it's not obvious (so error-prone) that the caller must pass volumes already filtered to createIsoDisksForConfigVolumes()

>sourceAndTargetPath() returning a tuple I think is error-prone, you need to check which one go first/second the source/target.

Sure but the caller just shot their own foot because they did not read the manual for the gun... In other words, it is as easy as to provide the wrong implementation to your defined interface.

I would at least suggest considering the change of `vmi` to `volumes`(not filtered) to prevent some crazy ideas and to help keep the focus of the function

[the remark this answers is not in the document]
Comment (germag) [reply pkg/config/config.go:195]: > 
> Sure but the caller just shot their own foot because they did not read the manual for the gun... 

:)

> In other words, it is as easy as to provide the wrong implementation to your defined interface.

Sure, but I think it's a bit more obvious

> I would at least suggest considering the change of `vmi` to `volumes`(not filtered) to prevent some crazy ideas and to help keep the focus of the function

In that case, we also need to pass the `vmi.Spec.Domain.Devices.Disks`and the `vmi.Status.VolumeStatus`

[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/config/config.go:195]: Disregard my comment :)

**Review by xpivarc — COMMENTED** [review]
/hold cancel
